### PR TITLE
Cache pickposition

### LINKED ISSUE #5117: Cache pickPosition
Orthographic projection calls `pickPosition` to determine the correct frustum width. `ScreenSpaceCameraController` also calls it in `pickGlobe`. Additionally app code may be calling it multiple times per frame. Since this is an expensive operations that causes a CPU/GPU sync, it may be good to cache the `pickPosition` results per frame, or over multiple frames.

### PR DESCRIPTION
**Pull request by ateshuseyin**
fixes #5117

### CODE REVIEW

**Comment by hpinkos** [comment]
Thanks @ateshuseyin!  It looks like you have a jsHint error:

```
Specs/Scene/SceneSpec.js
line 852  col 68  Unnecessary semicolon.
```

You can run jsHint locally with `npm run jsHint`

@lilleyse do you want to review this?

**Comment by lilleyse** [comment]
Yes, I should be able to review today.

File: Source/Scene/Scene.js [hunk 1]
```diff
@@ -3043,6 +3046,21 @@ define([
      * @exception {DeveloperError} Picking from the depth buffer is not supported. Check pickPositionSupported.
      */
     Scene.prototype.pickPosition = function(windowPosition, result) {
+
+        if(defined(windowPosition) && pickPositionCache.hasOwnProperty(windowPosition.toString())){
+            var cached = pickPositionCache[windowPosition.toString()];
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: You may want to add a comment somewhere that the cached result could be `undefined`. This way it is clear to whoever is reading that `hasOwnProperty` is used instead of just `defined`.

File: Source/Scene/Scene.js [hunk 2]
```diff
@@ -3053,6 +3071,9 @@ define([
             var cart = projection.unproject(result, scratchPickPositionCartographic);
             ellipsoid.cartographicToCartesian(cart, result);
         }
+
+        pickPositionCache[windowPosition.toString()] = defined(result) ? result.clone() : result;
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Same comment here about using `Cartesian2.clone(result)`

File: Specs/Scene/SceneSpec.js [hunk 3]
```diff
@@ -838,6 +838,77 @@ defineSuite([
...
+            var primitives = s.primitives;
+            primitives.add(rectanglePrimitive);
+        });
+
+        expect(s).toRenderAndCall(function() {
+            s.pickPosition(windowPosition);
+            expect(s.pickPositionWorldCoordinates.calls.count()).toEqual(2);
+
+            s.pickPosition(windowPosition);
+            expect(s.pickPositionWorldCoordinates.calls.count()).toEqual(2);
```
Comment (lilleyse) [inline Specs/Scene/SceneSpec.js]: It's fine to just do just one `expect(s.pickPositionWorldCoordinates)` check at the end.

File: Specs/Scene/SceneSpec.js [hunk 4]
```diff
@@ -838,6 +838,77 @@ defineSuite([
...
+        if (!s.pickPositionSupported) {
+            return;
+        }
+
+        var rectangle = Rectangle.fromDegrees(-100.0, 30.0, -90.0, 40.0);
+        s.camera.setView({ destination : rectangle });
+
+        var canvas = s.canvas;
+        var windowPosition = new Cartesian2(canvas.clientWidth / 2, canvas.clientHeight / 2);
+        spyOn(s, 'pickPositionWorldCoordinates').and.callThrough();;
```
Comment (lilleyse) [inline Specs/Scene/SceneSpec.js]: Extra `;` as @hpinkos mentioned.

File: Specs/Scene/SceneSpec.js [hunk 5]
```diff
@@ -838,6 +838,77 @@ defineSuite([
...
+        // s.useDepthPicking = true;
+
+        // expect(s).toRenderAndCall(function() {
+        //     var position = s.pickPosition(windowPosition);
+
+        //     expect(s.pickPositionWorldCoordinates).toHaveBeenCalled();
+        //     position = s.pickPosition(windowPosition);
+
+        //     expect(s.pickPositionWorldCoordinates.calls.count()).toEqual(1);
+        // });
```
Comment (lilleyse) [inline Specs/Scene/SceneSpec.js]: Remove commented code.

File: Source/Scene/Scene.js [hunk 6]
```diff
@@ -3024,7 +3025,9 @@ define([
 
     var scratchPickPositionCartographic = new Cartographic();
 
-    /**
+    var pickPositionCache={};
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Style: `var pickPositionCache = {};`

Comment (lilleyse) [inline Source/Scene/Scene.js]: In the case of multiple scenes, this may be better as a private variable of `Scene`.

File: Specs/Scene/SceneSpec.js (hunk 3 above)
Comment (ateshuseyin) [inline Specs/Scene/SceneSpec.js]: Second expect is necessary I think. First pickPosition is calculated and call count increment but at second result is returned from cache

**Comment by ateshuseyin** [comment]
fixed belong to review.

File: Source/Scene/Scene.js [hunk 7]
```diff
@@ -2523,6 +2525,8 @@ define([
     var scratchEyeTranslation = new Cartesian3();
 
     function render(scene, time) {
+        scene._pickPositionCache = {}; //clean <cache></cache>
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Fix comment.

File: Source/Scene/Scene.js [hunk 8]
```diff
@@ -2616,6 +2620,7 @@ define([
      * @private
      */
     Scene.prototype.render = function(time) {
+
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Remove empty line.

File: Source/Scene/Scene.js [hunk 9]
```diff
@@ -3024,7 +3029,7 @@ define([
 
     var scratchPickPositionCartographic = new Cartographic();
 
-    /**
+     /**
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Undo this change, the old formatting is consistent with the other comments.

File: Source/Scene/Scene.js [hunk 10]
```diff
@@ -3043,6 +3048,16 @@ define([
      * @exception {DeveloperError} Picking from the depth buffer is not supported. Check pickPositionSupported.
      */
     Scene.prototype.pickPosition = function(windowPosition, result) {
+        var cacheKey;
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: As mentioned before, the caching belongs in `pickPositionWorldCoordinates`. For example, both `Camera` and `ScreenSpaceCameraController` call it instead of `pickPosition`.

File: Source/Scene/Scene.js [hunk 11]
```diff
@@ -3043,6 +3048,16 @@ define([
...
+
+        if(defined(windowPosition)){
+            cacheKey = windowPosition.x + '-' + windowPosition.y;
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Why not use `toString` like before?

File: Source/Scene/Scene.js [hunk 12]
```diff
@@ -3043,6 +3048,16 @@ define([
...
+        }
+
+        if(defined(windowPosition) && this._pickPositionCache.hasOwnProperty(cacheKey)){
+            return Cartesian2.clone(this._pickPositionCache[cacheKey], result);
+        }
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: When this code is moved to `pickPositionWorldCoordinates` the `defined(windowPosition)` checks here and above won't be needed.

File: Source/Scene/Scene.js [hunk 13]
```diff
@@ -3043,6 +3048,16 @@ define([
      * @exception {DeveloperError} Picking from the depth buffer is not supported. Check pickPositionSupported.
      */
     Scene.prototype.pickPosition = function(windowPosition, result) {
+        var cacheKey;
+
+        if(defined(windowPosition)){
+            cacheKey = windowPosition.x + '-' + windowPosition.y;
+        }
+
+        if(defined(windowPosition) && this._pickPositionCache.hasOwnProperty(cacheKey)){
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: Style, add space: `if (defined(windowPosition)`

File: Specs/Scene/SceneSpec.js (hunk 3 above)
Comment (lilleyse) [inline Specs/Scene/SceneSpec.js]: One check should be fine because if the result is 2 then all the previous steps have used the cache correctly. But I'm also ok with keeping it as-is.

File: Specs/Scene/SceneSpec.js [hunk 14]
```diff
@@ -838,6 +838,43 @@ defineSuite([
...
+        });
+
+        expect(s).toRenderAndCall(function() {
+            s.pickPosition(windowPosition);
+            expect(s.pickPositionWorldCoordinates.calls.count()).toEqual(2);
+
+            s.pickPosition(windowPosition);
+            expect(s.pickPositionWorldCoordinates.calls.count()).toEqual(2);
+        });
+
```
Comment (lilleyse) [inline Specs/Scene/SceneSpec.js]: Remove empty line.

**Comment by pjcozzi** [comment]
Thanks for this contribution, @ateshuseyin!

@lilleyse please ping me when you think this is ready, I'd like to take a quick look before merging into master.

File: Specs/Scene/SceneSpec.js (hunk 14 above)
Comment (ateshuseyin) [inline Specs/Scene/SceneSpec.js]: First pickPosition calculates and returns undefined. The second one returns undefined from cache. Third calculates a valid Cartesian3 then forth one returns valid value from cache. So both calls are needed to test correctly. Because the flow of undefined and valid results are seperate.

**Comment by ateshuseyin** [comment]
Thanks cesium. It is great work.

I think all done.

File: Source/Scene/Scene.js [hunk 15]
```diff
@@ -2954,6 +2958,12 @@ define([
         }
         //>>includeEnd('debug');
 
+        var cacheKey;
```
Comment (lilleyse) [inline Source/Scene/Scene.js]: `cacheKey` needs to be set.

**Comment by ateshuseyin** [comment]
@lilleyse I think this is last commit :)

**Comment by lilleyse** [comment]
@pjcozzi this is ready for you to look at.

File: Source/Scene/Scene.js [hunk 16]
```diff
@@ -2548,6 +2550,8 @@ define([
     var scratchEyeTranslation = new Cartesian3();
 
     function render(scene, time) {
+        scene._pickPositionCache = {}; //clean cache
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: Can we avoid doing this allocation per frame?  We are really careful about not doing allocation per vertex, per command, per primitive, etc., but we even try to avoid per frame allocations.

Instead, perhaps set a flag like `_pickPositionCacheDirty` to `true`.  Then in `pickPositionWorldCoordinates`, clear the cache and set dirty to false if it was true.

**Comment by pjcozzi** [comment]
Just that comment.  Thanks again @ateshuseyin, nice job with the tests!

**Comment by ateshuseyin** [comment]
@pjcozzi done

**Comment by lilleyse** [comment]
This looks ready to me, @pjcozzi do you want to merge?

**Comment by pjcozzi** [comment]
Looks good, but are we sure about all the changes to the `.idea/*` files that came in with 435496178be223b5b9fb6742bfe5ea5ca3f02c5d?

**Comment by lilleyse** [comment]
I think those are related to the recent Webstorm update, which haven't been incorporated into 3d-tiles yet. I'll merge master into 3d-tiles first to be sure.

**Comment by lilleyse** [comment]
They aren't in master, so I just reverted those changes.

**Comment by pjcozzi** [comment]
Thanks @ateshuseyin @lilleyse!
